cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MessagingMetricsProcessorService.java

```diff
@@ -0,0 +1,307 @@
+          }
+        }
+      } catch (TopicNotFoundException | IOException e) {
+        LOG.error("Failed to fetch metrics records", e);
+      }
+
+      if (records.isEmpty()) {
+        LOG.info("No records to process.");
+        return;
+      }
+      persistRecords(false);
+    }
+
+    public void terminate() {
+      LOG.info("Terminate requested {}", getName());
+      if (!records.isEmpty()) {
+        persistRecords(true);
+      }
+      interrupt();
+    }
+
+    private void persistRecords(boolean terminating) {
+      if (records.size() <= fetcherPersistThreshold && !terminating) {
+        return;
+      }
+      try {
+        addProcessingStats(records);
+        metricStore.add(records);
+        LOG.info("Persisted {} metrics in MetricStore in thread {}", records.size(), this.getName());
+        // Persist lastMessageId when its content differs from INITIAL_LAST_MESSAGE_ID
+        if (!Arrays.equals(lastMessageId, INITIAL_LAST_MESSAGE_ID)) {
+          persistMessageId();
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("Failed to add metrics data to a store", e);
+      }
+
+      recordsProcessed += records.size();
+      // avoid logging more than once a minute
+      if (System.currentTimeMillis() > lastLoggedMillis + TimeUnit.MINUTES.toMillis(1)) {
+        lastLoggedMillis = System.currentTimeMillis();
+        LOG.info("{} metrics records processed in thread {}. Last record time: {}.",
+                 recordsProcessed, this.getName(), records.get(records.size() - 1).getTimestamp());
+      }
+      records.clear();
+    }
+
+    private void addProcessingStats(List<MetricValues> records) {
+      if (records.isEmpty()) {
+        return;
+      }
+      int count = records.size();
+      long now = System.currentTimeMillis();
+      long delay = now - TimeUnit.SECONDS.toMillis(records.get(records.size() - 1).getTimestamp());
+      records.add(
+        new MetricValues(metricsContextMap, TimeUnit.MILLISECONDS.toSeconds(now),
+                         ImmutableList.of(new MetricValue("metrics.process.count", MetricType.COUNTER, count),
+                                          new MetricValue("metrics.process.delay.ms", MetricType.GAUGE, delay))));
+    }
+
+    private void persistMessageId() {
+      try {
+        metaTable.save(topicIdMetaKey, lastMessageId);
+        LOG.info("Persisted last processed MessageId: {} in thread {}", lastMessageId, this.getName());
+      } catch (Exception e) {
+        // Simple log and ignore the error.
+        LOG.error("Failed to persist consumed messageId. {}", e.getMessage(), e);
+      }
+    }
+  }
+
+  private class TopicIdMetaKey implements MetricsMetaKey {
+    byte[] key;
+
+    TopicIdMetaKey(TopicId metricsTopic) {
+      this.key = MessagingUtils.toMetadataRowKey(metricsTopic);
+    }
+
+    @Override
+    public byte[] getKey() {
+      return key;
+    }
+  }
+
+  private class PayloadInputStream extends ByteArrayInputStream {
```

Comment:
It's just a wrapper class for ByteArrayInputStream, so that the internal `byte[]` can be reset and no need to create a new ByteArrayInputStream at every iteration. Made this change because of Terence's comment